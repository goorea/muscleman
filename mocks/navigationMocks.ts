export const navigationDispatchMock = jest.fn();
export const navigationNavigateMock = jest.fn();
export const navigationGoBackMock = jest.fn();
export const navigationSetOptionsMock = jest.fn();
export const navigationSetParamsMock = jest.fn();
export const navigationCanGoBackMock = jest.fn();
export const navigationGetParentMock = jest.fn();
export const navigationGetStateMock = jest.fn();
export const navigationReplaceMock = jest.fn();
export const navigationIsFocusedMock = jest.fn();
export const navigationResetMock = jest.fn();
export const navigationPopMock = jest.fn();
export const navigationPushMock = jest.fn();
export const navigationPopToTopMock = jest.fn();
export const navigationAddListenerMock = jest.fn();
export const navigationRemoveListenerMock = jest.fn();
export const navigationJumpToMock = jest.fn();
export const navigationMock = {
  dispatch: navigationDispatchMock,
  navigate: navigationNavigateMock,
  goBack: navigationGoBackMock,
  setOptions: navigationSetOptionsMock,
  setParams: navigationSetParamsMock,
  canGoBack: navigationCanGoBackMock,
  getParent: navigationGetParentMock,
  getState: navigationGetStateMock,
  isFocused: navigationIsFocusedMock,
  reset: navigationResetMock,
  addListener: navigationAddListenerMock,
  removeListener: navigationRemoveListenerMock,
  replace: navigationReplaceMock,
  pop: navigationPopMock,
  push: navigationPushMock,
  popToTop: navigationPopToTopMock,
  jumpTo: navigationJumpToMock,
};
