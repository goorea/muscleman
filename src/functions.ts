import StackFlashMessage, { Options } from 'react-native-stack-flash-message';

import { WeightSet } from '@src/types';
import { Set, TrainingType } from '@src/types/graphql';

export const flash = (options: Options) => StackFlashMessage.show(options);

export const getTrainingTypeForKorean = (type: TrainingType): string => {
  switch (type) {
    case TrainingType.Lower:
      return '하체';
    case TrainingType.Chest:
      return '가슴';
    case TrainingType.Back:
      return '등';
    case TrainingType.Shoulder:
      return '어깨';
    case TrainingType.Arm:
      return '팔';
    case TrainingType.Abdominal:
      return '복근';
    case TrainingType.Cardiovascular:
      return '유산소';
    default:
      return '기타';
  }
};

export const isWeightSets = (sets: Set[]): sets is WeightSet[] =>
  sets.every(
    set =>
      (set as WeightSet).weight !== undefined &&
      (set as WeightSet).count !== undefined,
  );
