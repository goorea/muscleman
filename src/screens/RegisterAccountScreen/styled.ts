import styled from 'styled-components/native';

import Button from '@src/components/Button';
import Text from '@src/components/Text';

export const Container = styled.View`
  padding: 20px;
`;

export const Title = styled(Text)`
  margin-bottom: 20px;
`;

export const Submit = styled(Button)`
  margin-top: auto;
`;
