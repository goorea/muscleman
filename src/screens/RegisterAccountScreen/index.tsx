import { CompositeScreenProps } from '@react-navigation/native';
import { NativeStackScreenProps } from '@react-navigation/native-stack';
import React from 'react';
import { useForm, Controller } from 'react-hook-form';

import KeyboardAvoidingScrollView from '@src/components/KeyboardAvoidingScrollView';
import {
  AuthStackParamList,
  RegisterStackParamList,
} from '@src/types/navigation';

import useRenders from './hooks/useRenders';
import useRules from './hooks/useRules';
import { Container, Title, Submit } from './styled';

type P = CompositeScreenProps<
  NativeStackScreenProps<RegisterStackParamList, 'RegisterAccount'>,
  NativeStackScreenProps<AuthStackParamList>
>;

export type RegisterAccountFormInput = {
  email: string;
  password: string;
  password_confirmation: string;
};

const RegisterAccountScreen: React.FC<P> = ({ navigation }) => {
  const { control, handleSubmit, getValues } =
    useForm<RegisterAccountFormInput>();
  const { emailRules, passwordRules, passwordConfirmationRules } =
    useRules(getValues);
  const onSubmit = (input: RegisterAccountFormInput) =>
    navigation.navigate('RegisterUser', input);
  const { emailRender, passwordRender, passwordConfirmationRender } =
    useRenders(handleSubmit(onSubmit));

  return (
    <KeyboardAvoidingScrollView
      floatChildren={<Submit onPress={handleSubmit(onSubmit)} title="다음" />}>
      <Container>
        <Title size={20} weight="bold">
          이메일과 비밀번호를 입력해주세요
        </Title>

        <Controller
          render={emailRender}
          name="email"
          control={control}
          rules={emailRules}
        />
        <Controller
          render={passwordRender}
          name="password"
          control={control}
          rules={passwordRules}
        />
        <Controller
          render={passwordConfirmationRender}
          name="password_confirmation"
          control={control}
          rules={passwordConfirmationRules}
        />
      </Container>
    </KeyboardAvoidingScrollView>
  );
};

export default RegisterAccountScreen;
