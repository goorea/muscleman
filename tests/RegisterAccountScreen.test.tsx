import { navigationMock, navigationNavigateMock } from '@mocks/navigationMocks';
import { act, fireEvent, render } from '@testing-library/react-native';
import React from 'react';

import RegisterAccountScreen from '@src/screens/RegisterAccountScreen';
import { wrapper } from '@tests/functions';

describe('RegisterAccountScreen 컴포넌트', () => {
  const rendered = () =>
    render(
      <RegisterAccountScreen
        navigation={navigationMock}
        route={{
          key: '',
          name: 'RegisterAccount',
        }}
      />,
      { wrapper },
    );

  it('렌더링이 올바르게 된다', () => {
    const { toJSON } = rendered();

    expect(toJSON()).toMatchSnapshot();
  });

  it('이메일 필드와 비밀번호 필드와 비밀번호 확인 필드와 다음 버튼이 보인다', () => {
    const { queryByTestId, queryByText } = rendered();

    expect(queryByTestId('emailField')).not.toBeNull();
    expect(queryByTestId('passwordField')).not.toBeNull();
    expect(queryByTestId('passwordConfirmationField')).not.toBeNull();
    expect(queryByText('다음')).not.toBeNull();
  });

  it('이메일 필드와 비밀번호 필드와 비밀번호 확인 필드를 비우고 다음 버튼을 누르면 에러 메세지들이 보인다', async () => {
    const { queryByText, getByText } = rendered();

    expect(queryByText('이메일을 입력해주세요')).toBeNull();
    expect(queryByText('비밀번호를 입력해주세요')).toBeNull();
    expect(queryByText('비밀번호 확인을 입력해주세요')).toBeNull();

    await act(async () => fireEvent.press(getByText('다음')));

    expect(queryByText('이메일을 입력해주세요')).not.toBeNull();
    expect(queryByText('비밀번호를 입력해주세요')).not.toBeNull();
    expect(queryByText('비밀번호 확인을 입력해주세요')).not.toBeNull();
  });

  it('이메일 필드는 이메일 형식이 아니라면 에러 메세지가 보인다', async () => {
    const { getByTestId, queryByText, getByText } = rendered();

    fireEvent.changeText(getByTestId('emailField'), 'Hello World');

    await act(async () => fireEvent.press(getByText('다음')));

    expect(queryByText('이메일 형식이 아닙니다')).not.toBeNull();
  });

  it('비밀번호 필드와 비밀번호 확인 필드는 8글자 이상이어야 한다', async () => {
    const { getByTestId, queryByText, getByText } = rendered();

    fireEvent.changeText(getByTestId('emailField'), 'john@example.com');
    fireEvent.changeText(getByTestId('passwordField'), 'a'.repeat(7));

    await act(async () => fireEvent.press(getByText('다음')));

    expect(queryByText('8글자보다 적습니다')).not.toBeNull();
  });

  it('올바른 값을 입력한 이후에는 RegisterUserScreen 페이지로 이동할 수 있다', async () => {
    const { getByTestId, getByText } = rendered();

    fireEvent.changeText(getByTestId('emailField'), 'john@example.com');
    fireEvent.changeText(getByTestId('passwordField'), 'a'.repeat(8));
    fireEvent.changeText(
      getByTestId('passwordConfirmationField'),
      'a'.repeat(8),
    );

    await act(async () => fireEvent.press(getByText('다음')));

    expect(navigationNavigateMock).toHaveBeenCalledWith(
      'RegisterUser',
      expect.anything(),
    );
  });
});
