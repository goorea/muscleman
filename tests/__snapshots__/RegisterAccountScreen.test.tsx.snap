// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RegisterAccountScreen 컴포넌트 렌더링이 올바르게 된다 1`] = `
<View
  onLayout={[Function]}
  style={
    Array [
      Array [
        Object {
          "flexBasis": 0,
          "flexGrow": 1,
          "flexShrink": 1,
        },
      ],
      Object {
        "paddingBottom": 0,
      },
    ]
  }
>
  <RCTScrollView>
    <View>
      <View
        style={
          Array [
            Object {
              "paddingBottom": 20,
              "paddingLeft": 20,
              "paddingRight": 20,
              "paddingTop": 20,
            },
          ]
        }
      >
        <Text
          size={20}
          style={
            Array [
              Object {
                "color": "#000000",
                "fontFamily": "SpoqaHanSansNeo-Bold",
                "fontSize": 20,
              },
              Object {
                "marginBottom": 20,
              },
            ]
          }
          weight="bold"
        >
          이메일과 비밀번호를 입력해주세요
        </Text>
        <View
          style={
            Array [
              Object {},
            ]
          }
        >
          <View
            style={
              Array [
                Object {
                  "alignItems": "baseline",
                  "borderBottomColor": "#bbb",
                  "borderBottomWidth": 1,
                  "marginBottom": 4,
                  "paddingBottom": 10,
                  "paddingLeft": 0,
                  "paddingRight": 0,
                  "paddingTop": 10,
                },
              ]
            }
          >
            <Text
              collapsable={false}
              color="foreground"
              nativeID="animatedComponent"
              onLayout={[Function]}
              style={
                Object {
                  "color": "#000000",
                  "fontFamily": "SpoqaHanSansNeo-Regular",
                  "fontSize": 16,
                  "transform": Array [
                    Object {
                      "translateX": 0,
                    },
                    Object {
                      "translateY": 16,
                    },
                    Object {
                      "scale": 1,
                    },
                  ],
                }
              }
            >
              이메일
            </Text>
            <TextInput
              autoCompleteType="email"
              autoFocus={true}
              keyboardType="email-address"
              label="이메일"
              name="email"
              onBlur={[Function]}
              onChange={[Function]}
              onChangeText={[Function]}
              onFocus={[Function]}
              onSubmitEditing={[Function]}
              returnKeyType="next"
              style={
                Array [
                  Object {
                    "color": "#000000",
                    "fontFamily": "SpoqaHanSansNeo-Regular",
                    "fontSize": 16,
                    "paddingBottom": 0,
                    "paddingLeft": 0,
                    "paddingRight": 0,
                    "paddingTop": 0,
                    "width": "100%",
                  },
                ]
              }
              testID="emailField"
            />
          </View>
        </View>
        <View
          style={
            Array [
              Object {},
            ]
          }
        >
          <View
            style={
              Array [
                Object {
                  "alignItems": "baseline",
                  "borderBottomColor": "#bbb",
                  "borderBottomWidth": 1,
                  "marginBottom": 4,
                  "paddingBottom": 10,
                  "paddingLeft": 0,
                  "paddingRight": 0,
                  "paddingTop": 10,
                },
              ]
            }
          >
            <Text
              collapsable={false}
              color="foreground"
              nativeID="animatedComponent"
              onLayout={[Function]}
              style={
                Object {
                  "color": "#000000",
                  "fontFamily": "SpoqaHanSansNeo-Regular",
                  "fontSize": 16,
                  "transform": Array [
                    Object {
                      "translateX": 0,
                    },
                    Object {
                      "translateY": 16,
                    },
                    Object {
                      "scale": 1,
                    },
                  ],
                }
              }
            >
              비밀번호
            </Text>
            <TextInput
              label="비밀번호"
              name="password"
              onBlur={[Function]}
              onChange={[Function]}
              onChangeText={[Function]}
              onFocus={[Function]}
              onSubmitEditing={[Function]}
              returnKeyType="next"
              secureTextEntry={true}
              style={
                Array [
                  Object {
                    "color": "#000000",
                    "fontFamily": "SpoqaHanSansNeo-Regular",
                    "fontSize": 16,
                    "paddingBottom": 0,
                    "paddingLeft": 0,
                    "paddingRight": 0,
                    "paddingTop": 0,
                    "width": "100%",
                  },
                ]
              }
              testID="passwordField"
            />
          </View>
        </View>
        <View
          style={
            Array [
              Object {},
            ]
          }
        >
          <View
            style={
              Array [
                Object {
                  "alignItems": "baseline",
                  "borderBottomColor": "#bbb",
                  "borderBottomWidth": 1,
                  "marginBottom": 4,
                  "paddingBottom": 10,
                  "paddingLeft": 0,
                  "paddingRight": 0,
                  "paddingTop": 10,
                },
              ]
            }
          >
            <Text
              collapsable={false}
              color="foreground"
              nativeID="animatedComponent"
              onLayout={[Function]}
              style={
                Object {
                  "color": "#000000",
                  "fontFamily": "SpoqaHanSansNeo-Regular",
                  "fontSize": 16,
                  "transform": Array [
                    Object {
                      "translateX": 0,
                    },
                    Object {
                      "translateY": 16,
                    },
                    Object {
                      "scale": 1,
                    },
                  ],
                }
              }
            >
              비밀번호 확인
            </Text>
            <TextInput
              label="비밀번호 확인"
              name="passwordConfirmation"
              onBlur={[Function]}
              onChange={[Function]}
              onChangeText={[Function]}
              onFocus={[Function]}
              onSubmitEditing={[Function]}
              returnKeyType="next"
              secureTextEntry={true}
              style={
                Array [
                  Object {
                    "color": "#000000",
                    "fontFamily": "SpoqaHanSansNeo-Regular",
                    "fontSize": 16,
                    "paddingBottom": 0,
                    "paddingLeft": 0,
                    "paddingRight": 0,
                    "paddingTop": 0,
                    "width": "100%",
                  },
                ]
              }
              testID="passwordConfirmationField"
            />
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    nativeID="animatedComponent"
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#2687ED",
        "justifyContent": "center",
        "marginTop": "auto",
        "opacity": 1,
        "paddingBottom": 14,
        "paddingLeft": 20,
        "paddingRight": 20,
        "paddingTop": 14,
      }
    }
  >
    <Text
      color="white"
      style={
        Array [
          Object {
            "color": "#ffffff",
            "fontFamily": "SpoqaHanSansNeo-Bold",
            "fontSize": 16,
          },
        ]
      }
      weight="bold"
    >
      다음
    </Text>
  </View>
</View>
`;
